src/coreclr/src/System.Private.CoreLib/src/System/Threading/Thread.CoreCLR.cs

```diff
@@ -486,19 +486,31 @@ private static int CalculateOptimalMaxSpinWaitsPerSpinIteration()
         [MethodImpl(MethodImplOptions.InternalCall)]
         private static extern int GetCurrentProcessorNumber();
 
-        // The upper bits of t_currentProcessorIdCache are the currentProcessorId. The lower bits of
-        // the t_currentProcessorIdCache are counting down to get it periodically refreshed.
-        // TODO: Consider flushing the currentProcessorIdCache on Wait operations or similar
-        // actions that are likely to result in changing the executing core
-        [ThreadStatic]
-        private static int t_currentProcessorIdCache;
+        // t_currentProcessorId lives in a separate class to make sure the class is fully initialized by the time we use the field
+        private class CoreIdCache
+        {
+            // The upper bits of t_currentProcessorId are the currentProcessorId. The lower bits of
+            // the t_currentProcessorIdCache are counting down to get it periodically refreshed.
+            // TODO: Consider flushing the currentProcessorIdCache on Wait operations or similar
+            // actions that are likely to result in changing the executing core
+            [ThreadStatic]
+            internal static int t_currentProcessorId;
+        }
 
         private const int ProcessorIdCacheShift = 16;
         private const int ProcessorIdCacheCountDownMask = (1 << ProcessorIdCacheShift) - 1;
-        private const int ProcessorIdRefreshRate = 5000;
+        // 50 is our best guess.
+        // Based on further calibration it is likley to be adjusted lower.
+        // In relatively rare cases of a slow GetCurrentProcessorNumber, it may be recalibrated to a higher number.
+        private static int ProcessorIdRefreshRate = 50;
```

Comment:
Coding conventions: static fields should start with `s_`, etc.

(multiple places)

Reply:
I keep forgetting about conventions when statics are constant-like. Will change.

---
In reply to: [352953273](https://github.com/dotnet/runtime/pull/467#discussion_r352953273) [](ancestors = 352953273)